### ISSUE TITLE
straight skeleton: faces which are not simple

### ISSUE THREAD

**Issue by khumarahn** [CGAL/cgal#5177]
Hi. I am working on an implementation of "roof" based on straight skeleton for openscad https://github.com/openscad/openscad/pull/3486

I am hitting a strange error, where a face of a straight skeleton is not simple. Below is a short code (modified example for straight skeleton with holes) which reproduces the issue:

```
#include<boost/shared_ptr.hpp>

#include<CGAL/Exact_predicates_inexact_constructions_kernel.h>
#include<CGAL/Polygon_with_holes_2.h>
#include<CGAL/create_straight_skeleton_from_polygon_with_holes_2.h>

#include "print.h"

typedef CGAL::Exact_predicates_inexact_constructions_kernel K ;

typedef K::Point_2                    Point ;
typedef CGAL::Polygon_2<K>            Polygon_2 ;
typedef CGAL::Polygon_with_holes_2<K> Polygon_with_holes ;
typedef CGAL::Straight_skeleton_2<K>  Ss ;

typedef boost::shared_ptr<Ss> SsPtr ;

int main()
{
	Polygon_2 outer ;

	outer.push_back(Point(6.0004730224609375, -1.500885009765625));
	outer.push_back(Point(6.0007781982421875, -1.500640869140625));
	outer.push_back(Point(6.0009613037109375, -1.5002899169921875));
	outer.push_back(Point(6.001007080078125, -1.5));
	outer.push_back(Point(6.0009613037109375, 1.5002899169921875));
	outer.push_back(Point(6.0007781982421875, 1.500640869140625));
	outer.push_back(Point(6.0003814697265625, 1.5009307861328125));
	outer.push_back(Point(6, 1.501007080078125));
	outer.push_back(Point(-6.0002899169921875, 1.5009613037109375));
	outer.push_back(Point(-6.000640869140625, 1.5007781982421875));
	outer.push_back(Point(-6.0009307861328125, 1.5003814697265625));
	outer.push_back(Point(-6.001007080078125, 1.5));
	outer.push_back(Point(-6.0009613037109375, -1.5002899169921875));
	outer.push_back(Point(-6.0007781982421875, -1.500640869140625));
	outer.push_back(Point(-6.0003814697265625, -1.5009307861328125));
	outer.push_back(Point(-6, -1.501007080078125));
	outer.push_back(Point(6, -1.501007080078125));

	Polygon_2 hole ;

	hole.push_back(Point(-4.998992919921875, -0.998992919921875));
	hole.push_back(Point(-4.998992919921875, 0.998992919921875));
	hole.push_back(Point(-0.10089111328125, 0.998992919921875));
	hole.push_back(Point(-0.0009307861328125, -0.0003814697265625));
	hole.push_back(Point(-0.0007781982421875, -0.000640869140625));
	hole.push_back(Point(-0.0003814697265625, -0.0009307861328125));
	hole.push_back(Point(0, -0.001007080078125));
	hole.push_back(Point(0.0004730224609375, -0.000885009765625));
	hole.push_back(Point(0.0007781982421875, -0.000640869140625));
	hole.push_back(Point(0.0009613037109375, -0.0002899169921875));
	hole.push_back(Point(0.10089111328125, 0.998992919921875));
	hole.push_back(Point(4.998992919921875, 0.998992919921875));
	hole.push_back(Point(4.998992919921875, 0.050933837890625));
	hole.push_back(Point(3.999755859375, 0.0009765625));
	hole.push_back(Point(3.99932861328125, 0.0007476806640625));
	hole.push_back(Point(3.9991455078125, 0.000518798828125));
	hole.push_back(Point(3.9990081787109375, 0.000152587890625));
	hole.push_back(Point(3.9990234375, -0.000244140625));
	hole.push_back(Point(3.9992523193359375, -0.00067138671875));
	hole.push_back(Point(3.999481201171875, -0.0008544921875));
	hole.push_back(Point(3.999847412109375, -0.0009918212890625));
	hole.push_back(Point(4.998992919921875, -0.050933837890625));
	hole.push_back(Point(4.998992919921875, -0.998992919921875));


	std::cout << "outer: "
		<< "is_convex: " << outer.is_convex()
		<< ", is_simple: " << outer.is_simple()
		<< ", is_clockwise_oriented: " << outer.is_clockwise_oriented()
		<< "\n";
	
	std::cout << "hole: "
		<< "is_convex: " << hole.is_convex()
		<< ", is_simple: " << hole.is_simple()
		<< ", is_clockwise_oriented: " << hole.is_clockwise_oriented()
		<< "\n";

	Polygon_with_holes poly( outer ) ;

	poly.add_hole( hole ) ;

	SsPtr iss = CGAL::create_interior_straight_skeleton_2(poly);

	int face_num = 0;
	for (auto face = iss->faces_begin(); face != iss->faces_end(); face++) {
		Polygon_2 face_poly;
		for (auto h = face->halfedge() ; ; ) {
			face_poly.push_back(h->vertex()->point());
			h = h->next();
			if (h == face->halfedge())
				break;
		}
		std::cout << "Face: " << face_num
			<< ". is_convex: " << face_poly.is_convex()
			<< ", is_simple: " << face_poly.is_simple();
		if (face_poly.is_simple()) {
			std::cout	<< ", is_clockwise_oriented: " << face_poly.is_clockwise_oriented();
		} else {
			std::cout << "\n" << face_poly;
		}
		std::cout	<< "\n";
		face_num++;
	}

	return 0;
}
```
Output:
```
outer: is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
hole: is_convex: 0, is_simple: 1, is_clockwise_oriented: 1
Face: 0. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 1. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 2. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 3. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 4. is_convex: 0, is_simple: 1, is_clockwise_oriented: 0
Face: 5. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 6. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 7. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 8. is_convex: 0, is_simple: 1, is_clockwise_oriented: 0
Face: 9. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 10. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 11. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 12. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 13. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 14. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 15. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 16. is_convex: 0, is_simple: 1, is_clockwise_oriented: 0
Face: 17. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 18. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 19. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 20. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 21. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 22. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 23. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 24. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 25. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 26. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 27. is_convex: 0, is_simple: 1, is_clockwise_oriented: 0
Face: 28. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 29. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 30. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 31. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 32. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 33. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 34. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 35. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 36. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 37. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 38. is_convex: 0, is_simple: 0
5 4.99899 -0.0509338 5.49999 0.425642 5.05254 2.7192e-07 4.00005 -5.33996e-06 3.99985 -0.000991821 
Face: 39. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
```
I am on gentoo linux, cgal version is 5.0.2. Also version 5.1 behaves the same.

**Comment by khumarahn** [comment]
I was too quick to click on "open" for this issue, it was actually a mouse clicking accident which I did not see how to revert. Nevertheless, I updated the problem description, hopefully it is clear what my problem is.

Face 38, the one which is not simple, is an antenna with 5 vertices:
```
5 4.99899 -0.0509338 5.49999 0.425642 5.05254 2.7192e-07 4.00005 -5.33996e-06 3.99985 -0.000991821
```

**Comment by khumarahn** [comment]
Here is an image of the contour:
![test](https://user-images.githubusercontent.com/5206045/99856981-1bf95500-2b82-11eb-8a48-ffa8eb3d87b1.png)
(what cannot be seen on the screenshot is that the corners are slightly rounded).

**Comment by khumarahn** [comment]
I tried with cgal-git on arch linux. The results are different: not there are two non-simple faces:
```
outer: is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
hole: is_convex: 0, is_simple: 1, is_clockwise_oriented: 1
Face: 0. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 1. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 2. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 3. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 4. is_convex: 0, is_simple: 1, is_clockwise_oriented: 0
Face: 5. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 6. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 7. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 8. is_convex: 0, is_simple: 1, is_clockwise_oriented: 0
Face: 9. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 10. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 11. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 12. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 13. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 14. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 15. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 16. is_convex: 0, is_simple: 1, is_clockwise_oriented: 0
Face: 17. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 18. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 19. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 20. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 21. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 22. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 23. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 24. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 25. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 26. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 27. is_convex: 0, is_simple: 0
6 0.100891 0.998993 -0.126256 1.24999 -1.07295e-06 1.11048 1.0678e-05 2.63755e-06 2.60396e-05 -1.05383e-05 0.000961304 -0.000289917 
Face: 28. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 29. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 30. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 31. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 32. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 33. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 34. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 35. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 36. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 37. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
Face: 38. is_convex: 0, is_simple: 0
5 4.99899 -0.0509338 5.49999 0.425642 5.05254 2.7192e-07 4.00005 -5.33996e-06 3.99985 -0.000991821 
Face: 39. is_convex: 1, is_simple: 1, is_clockwise_oriented: 0
```

**Comment by MaelRL** [comment]
The issue is fixed with https://github.com/CGAL/cgal/pull/5213. 

May I please use your polygon data in the CGAL testsuite?

**Comment by khumarahn** [comment]
Amazing, thanks a lot! Will this be a part of any upcoming release? I there a patch to say 5.1?

Of course you can use that polygon. Just in case, it is obtained in openscad by
```
        offset(0.001, $fn=64)
        difference() {
            square([12,3], center = true);
            polygon(points=[[-5,-1],[5,-1],[5,-0.05],[4,0],[5,0.05],[5,1],
                [0.1,1], [0,0], [-0.1,1], [-5,1]]);
        }
```

**Comment by MaelRL** [comment]
Thanks!

The pull request currently aims for the upcoming 5.2 release since, as you have seen, there have been a lot of improvements between 5.1 and 5.2. I will try to backport it once it has been tested, but if #4905 is required, then it will only be in 5.2.

**Comment by lrineau** [comment]
Fixed by https://github.com/CGAL/cgal/pull/5213.

**Comment by lrineau** [comment]
The fix will be in CGAL-5.2, released before the end of 2020.
